File: 0_3.leetcode/java/Divide_Two_Integers.java
public class Divide_Two_Integers {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }

        boolean isNegative = dividend < 0 ^ divisor < 0;
        dividend = Math.abs(dividend);
        divisor = Math.abs(divisor);
        int subQuotient = 0, quotient = 0;

        while (dividend - divisor >= 0) {
            for (subQuotient = 0; dividend - (divisor << subQuotient << 1) >= 0; subQuotient++);
            quotient += (1 << subQuotient);
            dividend -= (divisor << subQuotient);
        }

        return isNegative ? -quotient : quotient;
    }
}

File: 0_3.leetcode/kotlin/src/main/kotlin/Divide_Two_Integers.kt
import kotlin.math.abs

fun main() {
    val result = divide(2147483647, 2)
    print(result)
}

fun divide(dividend: Int, divisor: Int): Int {
    if (dividend == Int.MIN_VALUE && divisor == -1) {
        return Int.MAX_VALUE
    }

    val isNegative = (dividend < 0) xor (divisor < 0)
    var abDividend = abs(dividend)
    var abDivisor = abs(divisor)
    var quotient = 0

    while (abDividend - abDivisor >= 0) {
        var subQuotient = 0
        while (abDividend - (abDivisor shl subQuotient shl 1) >= 0) {
            subQuotient++
        }

        quotient += 1 shl subQuotient
        abDividend -= (abDivisor shl subQuotient)
    }

    return if (isNegative) -quotient else quotient
}